### ISSUE TITLE
Prevent calling a function with too many arguments

### ISSUE THREAD

**Issue by jakearchibald** [microsoft/TypeScript#42579]
# Suggestion

## 🔍 Search Terms

strict callback arguments, too many arguments warning, option

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I'd like an option to make the following invalid:

```ts
type Func = (arg1: number, arg2: number) => void;
const func: Func = (arg1: number) => undefined;
func(10, 11);
```

In this case, the second line should be invalid, as the provided function only takes one argument.

## 📃 Motivating Example

A real-world example:

```ts
const nextFrame = new Promise(requestAnimationFrame);
```

The above results in `requestAnimationFrame` being called with two arguments rather than the one it accepts. This is fragile, since, in future, a second argument may be added to `requestAnimationFrame` which makes the above code behave differently.

TypeScript could add an option (optional to maintain backwards compatibility) that would reject the above, since `requestAnimationFrame` has one parameter, whereas `new Promise` expects a callback that takes two arguments. This would protect the code from changes to `requestAnimationFrame` that would otherwise be considered backwards compatible.

## 💻 Use Cases

I wrote a blog post about the risks with this pattern, including a section on how TypeScript doesn't prevent it https://jakearchibald.com/2021/function-callback-risks/

**Comment by jakearchibald** [comment]
I don't think my proposed solution is right, as it'd cause TypeScript to reject things like this:

```ts
const doubler = (n: number) => n * 2;
const mappedArray = numberArray.map(doubler);
const nextFrame = new Promise(requestAnimationFrame);
```

Line 2 is fine, because `doubler` is designed to be a callback to `array.map`, whereas line 3 is bad because `requestAnimationFrame` isn't designed to be a callback to `new Promise`. However, in both cases there's a function being called with an incorrect number of arguments.

Maybe this isn't a typing problem, but I'll leave this open in case anyone thinks of a better solution.

**Comment by jakearchibald** [comment]
One way to hack around this problem right now is:

```ts
function doubleNumber(num: number, _: never) {
  return num * 2;
}
```

[This prevents](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/GYVwdgxgLglg9mABAEziARgGwKYDkQC262ATgBRiEBcilRpANIgPo1jYBupAlIgN4AoRIhLYoIEkjqIAVIgBMAbgEBfAQIDaARibymAZgC6AOgIBDAA5lUGHPnoluioA):

```ts
[1, 2, 3].map(doubleNumber);
```

…but it makes for a messy auto-complete, as the 'never' arg shows up. Maybe the solution is a more semantic way to mark a function as "cannot be assigned to a function type with more parameters". This would allow library authors to make this assertion, if they want to reserve additional params for future use, and this assertion could be added to DOM functions like `requestAnimationFrame`.

**Comment by MartinJohns** [comment]
[Why are functions with fewer parameters assignable to functions that take more parameters?](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-with-fewer-parameters-assignable-to-functions-that-take-more-parameters)

> This is the expected and desired behavior. [...]

---

> In this case, the second line should be invalid, as the provided function only takes one argument.

But the **type** clearly says the function takes two arguments. The implementation is irrelevant; the contract is important.

**Comment by jakearchibald** [comment]
> It most definitely would be a massive breaking change.

Please note the proposal was for an option, so it wouldn't have been a breaking change. Also see the follow-up proposal https://github.com/microsoft/TypeScript/issues/42579#issuecomment-770935178.

**Comment by jakearchibald** [comment]
@MartinJohns I have updated the OP with my follow-up suggestion, which doesn't depend on an option, and maintains backwards compatibility.

**Comment by intrnl** [comment]
Wouldn't it be more wise to make it opt-out instead?

It's a breaking change for sure if it's opt-out, but I think something like this is important, because we don't want APIs like `requestAnimationFrame` to be held back just because some people decided to use it improperly...

Library devs might not be worried by this due to semantic versioning, but it's still relieving that another parameter could just be added to a function without it necessarily being a "breaking change"

**Comment by fernap3** [comment]
Any thoughts on how an author would opt in to this behavior? A `strict` keyword with the function declaration?

**Comment by jakearchibald** [comment]
@fernap3 yeah, a `strict` or a `const` somewhere feels ok, but I don't have strong feelings.

**Comment by RyanCavanaugh** [comment]
> The above results in requestAnimationFrame being called with two arguments rather than the one it accepts. This is fragile, since, in future, a second argument may be added to requestAnimationFrame which makes the above code behave differently.

Historic examples of this actually happening would be useful. The relevant spec committees are aware of this behavior of JS and realize when they're making breaking changes like this.

**Comment by domenic** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=1164959 is one.

**Comment by jakearchibald** [comment]
@RyanCavanaugh I'll see if there are more standards examples, but here are some anecdotes from the Twitter thread:

https://twitter.com/gnapse/status/1355305922353102851
https://twitter.com/tabatkins/status/1355191322932510720
https://twitter.com/helmers/status/1355218756633554945

There's also [this lint rule](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-callback-reference.md), which is on-by-default in a popular eslint rule set.

I want to pick up on [something you said on Twitter](https://twitter.com/SeaRyanC/status/1356358711862730752):

> adding a parameter/argument to a function/callback invocation _is_ a breaking change, and having TS error because something might be subjected to a breaking change in the future means it would have to error on pretty much everything?

What amounts to a breaking change is a grey area. You could change a comment and break code that was relying on the `toString()` of the function, but I think the vast majority consider a comment change to be backwards compatible. Similarly, I think most folks consider this to be a backwards compatible change:

```ts
// v1.0.0
export function increment(num: number): number {
  return num + 1;
}
// v1.1.0
export function increment(num: number, by: number = 1): number {
  return num + by;
}
```

It's largely considered backwards compatible because `increment(5)` gives the same result in both cases. It isn't expected that folks will call the v1.0.0 version with `increment(5, 10)`. The eslint rule linked above refers to this kind of change as something you'd make in a "minor version".

**Comment by RyanCavanaugh** [comment]
> most folks consider this to be a backwards compatible change:

I don't know what to say except that those people are, factually, wrong. If a library author wants to keep the door open for adding parameters in the future in a way that isn't a breaking change, they need to `throw` today if the incoming arity is too high.

By way of analogy, most library authors don't treat "adding" a return value (e.g. going from implicit `undefined` to `number`) to be a breaking change, but if tomorrow `console.log` starts returning the number of characters printed like `Array#push` does, then this will break code like `foo(console.log(x) || x)`. TypeScript stops you from doing that because we think the `console.log` return value is *incidentally* `undefined` (i.e., `void`) rather than *intentionally* `undefined`, and we get bug reports about this because people think it's safe to rely on that function always producing a falsy value, and I think on balance they're correct and we should change `console.log`'s return type to be `undefined` because that's what's in the spec today and, barring evidence to the contrary, the spec for any given function isn't likely to change.

Ultimately the entire surface API of anything is subject to Hyrum's law and I'm struggling to draw a bright line where things "could plausibly change" and "couldn't plausibly change" in a way that doesn't require arguing about thousands of different functions in the DOM and core lib to represent this. Do we think `Array#forEach`'s callback could ever gain an additional arg? Why / why not? How do I apply that logic elsewhere?

**Comment by jakearchibald** [comment]
If that's TypeScript's position on this stuff, I'm happy for this issue to close as WONTFIX.

**Comment by RyanCavanaugh** [comment]
I'm happy to hear feedback on this; it's not clearly out-of-scope. If there are e.g. functions *likely* to acquire new params, we could mark those and then hide them in intellisense, for example.

**Comment by jakearchibald** [comment]
I'm exploring this further in https://github.com/heycam/webidl/issues/954. If we decide that we shouldn't try to mitigate this at runtime, then this issue should probably close too.

**Comment by lazytype** [comment]
I'm surprised I haven't seen anyone point this out, but it is fairly straight-forward to define the kind of functions you want, though possibly not as ergonomic desirable
```ts
export function increment(...[num]: [number]): number {
  return num + 1;
}

// okay
increment(42);

// type error
increment(42, 5)
```

In fact parameters of `Generator.prototype.next` is defined this way
